### PR TITLE
Delete pthread keys when the jemalloc library is destructed

### PR DESCRIPTION
**Pull request by ggouaillardet**
Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/53)

### CODE REVIEW

**Comment by ggouaillardet** [comment]
Refs. open-mpi/ompi#5206

**Comment by ggouaillardet** [comment]
A simple reproducer is in my ggouaillardet/memkindtest repository.

Basically, the key destructor is invoked __after__ `libmemkind.so` is unmmap'ed and leads to a crash.
A real life example is when the Open MPI Java bindings are used (and Open MPI is __not__ configure'd with `--disable-dlopen`)

I only fixed the issue for my environment (e.g. an up-to-date CentOS 7 box) and other environments (for example, but not limited to, Windows) have not been tested.

Bottom line, this PR is to be seen as a proof of concept and should be carefully reviewed.

**Comment by lukaszstolarczuk** [comment]
This PR is stale for a while - I'm allowing myself to close it. To be continued, details saved in issue: https://github.com/memkind/memkind/issues/874
